File: app/src/main/java/com/example/user/eventest/DateTimeDialog.java
package com.example.user.eventest;

import android.app.Dialog;
import android.content.DialogInterface;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.app.DialogFragment;
import android.support.v7.app.AlertDialog;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.DatePicker;
import android.widget.TimePicker;

import com.example.user.eventest.eventbus.events.DatePickerUpdateEvent;

import org.greenrobot.eventbus.EventBus;

import java.text.DateFormat;
import java.text.ParseException;
import java.util.Calendar;
import java.util.Date;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.Unbinder;

/**
 * Created by DR
 * on 25.12.2017.
 */

public class DateTimeDialog extends DialogFragment {
    private Date mDate;
    boolean isDatePickerShown = false;
    @BindView(R.id.datePicker)
    DatePicker datePicker;
    @BindView(R.id.timePicker)
    TimePicker timePicker;
    @BindView(R.id.fabDateTime)
    FloatingActionButton fabDateTime;
    private Unbinder unbinder;


    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        Bundle bundle = getArguments();
        String date = "";
        String time = "";
        if (bundle != null) {
            date = bundle.getString("date", "");
            time = bundle.getString("time", "");
        }
        DateFormat dateFormat = DateFormat.getDateTimeInstance(DateFormat.SHORT, DateFormat.SHORT);
        try {
            mDate = dateFormat.parse(date + " " + time);
        } catch (ParseException e) {
            e.printStackTrace();
        }

        final Calendar calendarDate = Calendar.getInstance();
        calendarDate.setTime(mDate);
        LayoutInflater li = LayoutInflater.from(getActivity());
        View dateTimeLayout = li.inflate(R.layout.date_time_picker, null);
        unbinder = ButterKnife.bind(this, dateTimeLayout);

        updateDatePicker(calendarDate);
        updateTimePicker(calendarDate);

        AlertDialog.Builder alertDialogBuilder = new AlertDialog.Builder(getActivity());
        alertDialogBuilder
                .setView(dateTimeLayout)
                .setCancelable(false)
                .setPositiveButton(R.string.OK,
                        new DialogInterface.OnClickListener() {
                            public void onClick(DialogInterface dialog, int id) {
                                Calendar calendar = Calendar.getInstance();
                                int hour;
                                int minute;
                                if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
                                    hour = timePicker.getCurrentHour();
                                    minute = timePicker.getCurrentMinute();
                                } else {
                                    hour = timePicker.getHour();
                                    minute = timePicker.getMinute();
                                }
                                calendar.set(datePicker.getYear(),
                                        datePicker.getMonth(),
                                        datePicker.getDayOfMonth(),
                                        hour, minute);
                                EventBus.getDefault().post(new DatePickerUpdateEvent(calendar));
                            }
                        })
                .setNegativeButton(R.string.cancel,
                        new DialogInterface.OnClickListener() {
                            public void onClick(DialogInterface dialog, int id) {
                                dialog.cancel();
                            }
                        });
        AlertDialog alertDialog = alertDialogBuilder.create();
        alertDialog.getWindow().getAttributes().windowAnimations = R.style.DateTimeDialogTheme;

        return alertDialog;

    }

    @OnClick(R.id.fabDateTime)
    void switchPickers() {
        if (isDatePickerShown) {
            datePicker.setVisibility(View.INVISIBLE);
            timePicker.setVisibility(View.VISIBLE);
            fabDateTime.setImageResource(
                    R.drawable.ic_date_range_white_24px);
        } else {
            datePicker.setVisibility(View.VISIBLE);
            timePicker.setVisibility(View.INVISIBLE);
            fabDateTime.setImageResource(
                    R.drawable.ic_access_time_white_24px);
        }
        isDatePickerShown = !isDatePickerShown;
    }

    private void updateTimePicker(Calendar c) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
            timePicker.setCurrentHour(c.get(Calendar.HOUR));
            timePicker.setCurrentMinute(c.get(Calendar.MINUTE));
        } else {
            timePicker.setHour(c.get(Calendar.HOUR));
            timePicker.setMinute(c.get(Calendar.MINUTE));
        }
    }

    private void updateDatePicker(Calendar c) {
        datePicker.updateDate(
                c.get(Calendar.YEAR),
                c.get(Calendar.MONTH),
                c.get(Calendar.DAY_OF_MONTH));
    }

    @Override
    public void onResume() {
        super.onResume();
    }


    @Override
    public void onDestroyView() {
        super.onDestroyView();
        unbinder.unbind();
    }

}

File: app/src/main/java/com/example/user/eventest/DateTimeDialog.kt
package com.example.user.eventest

import android.app.Dialog
import android.content.res.Configuration
import android.os.Build
import android.os.Bundle
import android.support.v4.app.DialogFragment
import android.support.v7.app.AlertDialog
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.example.user.eventest.eventbus.events.DatePickerUpdateEvent
import kotlinx.android.synthetic.*
import kotlinx.android.synthetic.main.date_time_picker.view.*
import org.greenrobot.eventbus.EventBus
import java.text.DateFormat
import java.text.ParseException
import java.util.*
import kotlin.properties.Delegates

/**
 * Created by DR
 * on 25.12.2017.
 */

class DateTimeDialog : DialogFragment() {
    private var mDate: Date = Date()
    private var isDatePickerShown = false
    private var dateTimeLayout: View by Delegates.notNull()

    @Override
    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val bundle = arguments
        var date = ""
        var time = ""
        if (bundle != null) {
            date = bundle.getString(TV_DATE_KEY, "")
            time = bundle.getString(TV_TIME_KEY, "")
        }
        val dateFormat = DateFormat.getDateTimeInstance(DateFormat.SHORT, DateFormat.SHORT)
        try {
            mDate = dateFormat.parse("$date $time")
        } catch (e: ParseException) {
            e.printStackTrace()
        }
        val calendarDate = Calendar.getInstance()
        calendarDate.time = mDate
        val li = LayoutInflater.from(activity)
        dateTimeLayout = li.inflate(R.layout.date_time_picker, null)

        updateDatePicker(calendarDate)
        updateTimePicker(calendarDate)
        repaintPickers()

        val alertDialogBuilder: AlertDialog.Builder = AlertDialog.Builder(dateTimeLayout.context)
        alertDialogBuilder
                .setView(dateTimeLayout)
                .setCancelable(false)
        val alertDialog = alertDialogBuilder.create()
        alertDialog.window.attributes.windowAnimations = R.style.DateTimeDialogTheme
        dateTimeLayout.btnOK.setOnClickListener {
            val calendar = getCalendarFromPickers()
            EventBus.getDefault().post(DatePickerUpdateEvent(calendar))
            dialog.cancel()
        }
        dateTimeLayout.btnCancel.setOnClickListener { dialog.cancel() }
        dateTimeLayout.fabDateTime.setOnClickListener {
            isDatePickerShown = !isDatePickerShown
            repaintPickers()
        }
        return alertDialog
    }

    @Suppress("DEPRECATION")
    private fun getCalendarFromPickers(): Calendar {
        val hour: Int
        val minute: Int
        val calendar = Calendar.getInstance()
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
            hour = dateTimeLayout.timePicker.currentHour
            minute = dateTimeLayout.timePicker.currentMinute
        } else {
            hour = dateTimeLayout.timePicker.hour
            minute = dateTimeLayout.timePicker.minute
        }
        calendar.set(dateTimeLayout.datePicker.year,
                dateTimeLayout.datePicker.month,
                dateTimeLayout.datePicker.dayOfMonth,
                hour, minute)
        return calendar
    }

    private fun repaintPickers() {
        clearFindViewByIdCache()
        if (isDatePickerShown) {
            dateTimeLayout.datePicker.visibility = View.VISIBLE
            dateTimeLayout.timePicker.visibility = View.INVISIBLE
            dateTimeLayout.fabDateTime.setImageResource(
                    R.drawable.ic_access_time_white_24px)
            dateTimeLayout.tvNow.text = DateFormat.getTimeInstance(DateFormat.SHORT)
                    .format(getCalendarFromPickers().time)
        } else {
            dateTimeLayout.datePicker.visibility = View.INVISIBLE
            dateTimeLayout.timePicker.visibility = View.VISIBLE
            dateTimeLayout.fabDateTime.setImageResource(
                    R.drawable.ic_date_range_white_24px)
            dateTimeLayout.tvNow.text = DateFormat.getDateInstance(DateFormat.SHORT)
                    .format(getCalendarFromPickers().time)
        }

    }

    @Suppress("DEPRECATION")
    private fun updateTimePicker(c: Calendar) {
        clearFindViewByIdCache()
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
            dateTimeLayout.timePicker.currentHour = c[Calendar.HOUR]
            dateTimeLayout.timePicker.currentMinute = c[Calendar.MINUTE]
        } else {
            dateTimeLayout.timePicker.hour = c[Calendar.HOUR]
            dateTimeLayout.timePicker.minute = c[Calendar.MINUTE]
        }
    }

    private fun updateDatePicker(c: Calendar) {
        clearFindViewByIdCache()
        dateTimeLayout.datePicker.updateDate(
                c[Calendar.YEAR],
                c[Calendar.MONTH],
                c[Calendar.DAY_OF_MONTH])
    }

    override fun onStart() {
        super.onStart()
        val dialog: Dialog = dialog
        val orientation = resources.configuration.orientation
        if (orientation == Configuration.ORIENTATION_LANDSCAPE) {
            dialog.window.setLayout(ViewGroup.LayoutParams.MATCH_PARENT,
                    ViewGroup.LayoutParams.MATCH_PARENT)
        }
    }
}

